keps/407-gang-scheduling/README.md

```diff
@@ -0,0 +1,407 @@
+# KEP-407: Gang Scheduling
+
+<!--
+This is the title of your KEP. Keep it short, simple, and descriptive. A good
+title can help communicate what the KEP is and should be considered as part of
+any review.
+-->
+
+<!--
+A table of contents is helpful for quickly jumping to sections of a KEP and for
+highlighting any additional information provided beyond the standard KEP
+template.
+
+Ensure the TOC is wrapped with
+  <code>&lt;!-- toc --&rt;&lt;!-- /toc --&rt;</code>
+tags, and then generate with `hack/update-toc.sh`.
+-->
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Interface](#interface)
+  - [Sequence Diagram](#sequence-diagram)
+  - [Pod Webhook](#pod-webhook)
+  - [Pod Controller](#pod-controller)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+      - [Unit tests](#unit-tests)
+      - [Integration tests](#integration-tests)
+      - [e2e tests](#e2e-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+This KEP aims to enable gang scheduling support for LeaderWorkerSet by creating a PodGroup for each replica.
```

Comment:
I think, it is better to mention this https://github.com/kubernetes/enhancements/issues/4671 in this KEP.

Reply:
Native `PodGroup` resource in Kubernetes would clear away the in-tree custom scheduler implementation necessity.

Reply:
As far as I know, this KEP has been inactive for a long time. Volcano, coscheduling scheduler plugin, and Yunikorn are currently the most mature gang schedulers. Integrating a custom scheduler can help to expand the lws ecosystem more quickly, especially during this popular period for inference. It's uncertain when this KEP will be merged, and I believe that integrating custom schedulers does not conflict with this KEP. It can be added later as the default implementation :)